### PR TITLE
Legolas - DolittleClient, Tenants and Resources

### PR DESCRIPTION
**Pull request by woksin**
## Summary

Adds the ability to get the configured tenants, and a MongoDB database per tenant from the Runtime through the client. Also renames the Client to DolittleClient, and introduces interfaces many places to simplify creation of mocks for testing purposes.

### Added

- ITenants tenants property on the client for getting all tenants from the Runtime
- IResourcesBuilder resources property on the client for getting resources for a specific tenant. Currently supports MongoDB. 
- IDolittleClient interface that DolittleClient implements, and interfaces for other classes in the client structure

### Changed

- Client renamed to DolittleClient
- ClientBuilder renamed to DolittleClientBuilder

### Fixed

- A bug in the implementation of `Cancellation` and `CancellationSource` where `instanceof Cancellation` does not work.
- A small typo in the exception thrown when the client cannot connect to the configured Runtime.

### CODE REVIEW

File: Source/resources/Resource.ts [hunk 1]
```diff
@@ -0,0 +1,77 @@
+// Copyright (c) Dolittle. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+import { CallRequestContext } from '@dolittle/contracts/Services/CallContext_pb';
```
Comment (jakhog) [inline Source/resources/Resource.ts:3]: A little empty line here

File: Source/resources/Resource.ts [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+
+/**
+ * Represents the base implementation of a {@link IResource}.
+ */
+export abstract class Resource<TRequest, TResponse extends ResponseLike> extends IResource {
+    /**
+     * Initializes an instance of the {@link Resource} class.
+     * @param tenant The tenant id.
+     * @param _client The resources client.
+     * @param _executionContext The execution context.
```
Comment (jakhog) [inline Source/resources/Resource.ts:26]: Other type of inconsistent docs

File: Source/resources/ResourceName.ts [hunk 3]
```diff
@@ -0,0 +1,30 @@
+// Copyright (c) Dolittle. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+import { ConceptAs } from '@dolittle/concepts';
+
+export type ResourceNameLike = string | ResourceName;
```
Comment (jakhog) [inline Source/resources/ResourceName.ts:6]: Docs

File: Source/resources/for_Dummy/when_nothing.ts [hunk 4]
```diff
@@ -0,0 +1,8 @@
+// Copyright (c) Dolittle. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+import { describeThis } from '@dolittle/typescript.testing';
+
+describeThis(__filename, () => {
```
Comment (jakhog) [inline Source/resources/for_Dummy/when_nothing.ts:6]: Have we changed all the tests to use the new `describeThis`?

File: Source/resources/mongoDB/MongoDBConnectionString.ts [hunk 5]
```diff
@@ -0,0 +1,29 @@
+// Copyright (c) Dolittle. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+import { ConceptAs } from '@dolittle/concepts';
+
+export type MongoDBConnectionStringLike = string | MongoDBConnectionString;
```
Comment (jakhog) [inline Source/resources/mongoDB/MongoDBConnectionString.ts:5]: Empty line and docs

File: Source/resources/mongoDB/IMongoDBResource.ts [hunk 6]
```diff
@@ -0,0 +1,17 @@
...
+
+/**
+ * Represents an implementation of {@link Resource} for MongoDB.
+ */
+export abstract class IMongoDBResource extends Resource<GetRequest, GetMongoDbResponse> {
+    /**
+     * Gets the the MongoDB connection string.
+     * @param {Cancellation} [cancellation] The optional {@link Cancellation}.
+     */
+    abstract getConnectionString(cancellation?: Cancellation): Promise<MongoDBConnectionString>;
```
Comment (jakhog) [inline Source/resources/mongoDB/IMongoDBResource.ts:16]: Is it going to be annoying for people to use the ConceptAs type from this API?

File: Source/resources/mongoDB/index.ts [hunk 7]
```diff
@@ -0,0 +1,6 @@
+// Copyright (c) Dolittle. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+export { MongoDBConnectionString } from './MongoDBConnectionString';
+export { IMongoDBResource } from './IMongoDBResource';
+export * as internal from './internal';
```
Comment (jakhog) [inline Source/resources/mongoDB/index.ts:6]: Do we usually export the internals like this? They aren't really ment to be used externally...

File: Source/resources/mongoDB/internal/MongoDBResource.ts [hunk 8]
```diff
@@ -0,0 +1,51 @@
...
+        return MongoDBConnectionString.from(response.getConnectionstring());
+    }
+
+    /** @inheritdoc */
+    protected createRequest(): GetRequest {
+        const request = new GetRequest();
+        request.setCallcontext(this.createCallContext());
+        return request;
+    }
+}
```
Comment (jakhog) [inline Source/resources/mongoDB/internal/MongoDBResource.ts:51]: Newline

Comment (jakhog) [reply Source/resources/mongoDB/internal/MongoDBResource.ts:51]: Maybe I can make the linter do all of these checks for me...

File: Source/resources/package.json [hunk 9]
```diff
@@ -0,0 +1,40 @@
...
+        "build": "tsc -b",
+        "clean": "tsc -b --clean",
+        "test": "mocha",
+        "lint": "eslint --quiet --ext .ts ./",
+        "lint:fix": "eslint --quiet --ext .ts ./ --fix"
+    },
+    "dependencies": {
+        "@dolittle/concepts": "5.0.1",
+        "@dolittle/rudiments": "5.0.1",
+        "@dolittle/runtime.contracts": "6.3.0-legolas.3",
```
Comment (jakhog) [inline Source/resources/package.json:36]: Upgrade contracts.

File: Source/tenancy/Tenant.ts [hunk 10]
```diff
@@ -0,0 +1,17 @@
...
+
+import { TenantId } from '@dolittle/sdk.execution';
+
+/**
+ * Represents a tenant.
+ */
+export class Tenant {
+    /**
+     * Initializes a new instance of {@link Tenant}.
+     *
```
Comment (jakhog) [inline Source/tenancy/Tenant.ts:12]: Extra line here

File: Source/tenancy/index.ts [hunk 11]
```diff
@@ -0,0 +1,8 @@
+// Copyright (c) Dolittle. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+export { FailedToGetAllTenants } from './FailedToGetAllTenants';
+export { ITenants } from './ITenants';
+export { Tenant } from './Tenant';
+
+export * as internal from './internal';
```
Comment (jakhog) [inline Source/tenancy/index.ts:8]: Again, internals exposed...

File: Source/tenancy/internal/Tenants.ts [hunk 12]
```diff
@@ -0,0 +1,45 @@
...
+import { TenantsClient } from '@dolittle/runtime.contracts/Tenancy/Tenants_grpc_pb';
+import { GetAllRequest, Tenant as PbTenant } from '@dolittle/runtime.contracts/Tenancy/Tenants_pb';
+import { TenantId } from '@dolittle/sdk.execution';;
+import { Cancellation } from '@dolittle/sdk.resilience';
+import { reactiveUnary } from '@dolittle/sdk.services';
+import { Logger } from 'winston';
+import { Tenant, ITenants, FailedToGetAllTenants } from '../index';
+
+/**
+ * Represents a client for Tenants and an implementation of @link } that knows how to register Event Types with the Runtime.
```
Comment (jakhog) [inline Source/tenancy/internal/Tenants.ts:13]: Broken `@link` thing.

**Review by jakhog — COMMENTED** [review]
Mostly minor things I'll fix - will leave the unanswered questions for us to do together. Also found something strange in `Source/embeddings/Embedding.ts` with the super call - I think maybe there is something strange going on with the hierarchy of interfaces etc...

File: Source/embeddings/for_Dummy/when_nothing.ts [hunk 13]
```diff
@@ -4,5 +4,5 @@
 import { describeThis } from '@dolittle/typescript.testing';
 
 describeThis(__filename, () => {
-    it('should be fine', () => true.should.be.true);
+    it('should be fine, mkay', () => true.should.be.true);
```
[the remark this answers is not in the document]
Comment (jakhog) [reply Source/embeddings/for_Dummy/when_nothing.ts:7]: It's gone - we don't need dummy tests anymore.

File: Source/resources/.eslintignore [hunk 14]
```diff
@@ -0,0 +1,3 @@
+node_modules
+Distribution
+wallaby.conf.js
```
[the remark this answers is not in the document]
Comment (jakhog) [reply Source/resources/.eslintignore:3]: Done

File: Source/resources/for_Dummy/when_nothing.ts (hunk 4 above)
Comment (jakhog) [reply Source/resources/for_Dummy/when_nothing.ts:6]: Yes we have :)

File: Source/resources/Resource.ts (hunk 1 above)
Comment (jakhog) [reply Source/resources/Resource.ts:3]: Done, with new rule.

File: Source/resources/mongoDB/internal/MongoDBResource.ts (hunk 8 above)
Comment (jakhog) [reply Source/resources/mongoDB/internal/MongoDBResource.ts]: It is now checking and fixed

File: Source/resources/mongoDB/index.ts (hunk 7 above)
Comment (jakhog) [reply Source/resources/mongoDB/index.ts:6]: Seems like we need it for now - so let's just leave it.

File: Source/tenancy/index.ts (hunk 11 above)
Comment (jakhog) [reply Source/tenancy/index.ts:8]: Seems like we need it for now - so let's just leave it.

File: Source/resources/ResourceName.ts (hunk 3 above)
Comment (jakhog) [reply Source/resources/ResourceName.ts:6]: Fixed with new rules.

File: Source/resources/mongoDB/MongoDBConnectionString.ts (hunk 5 above)
Comment (jakhog) [reply Source/resources/mongoDB/MongoDBConnectionString.ts:6]: Fixed with rules.

File: Source/tenancy/Tenant.ts (hunk 10 above)
Comment (jakhog) [reply Source/tenancy/Tenant.ts:12]: Fixed.

File: Source/resources/mongoDB/IMongoDBResource.ts (hunk 6 above)
Comment (woksin) [reply Source/resources/mongoDB/IMongoDBResource.ts:18]: I'm not sure. I think that I would have liked it if I were the consumer, but I'm also biased 😋

**Pull request merged**